arch/arc/core/mpu/arc_mpu.c

```diff
@@ -72,658 +40,21 @@ static inline u32_t _get_region_attr_by_type(u32_t type)
-{
-#if CONFIG_ARC_MPU_VER == 2
-	return _get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION) + 1;
-#elif CONFIG_ARC_MPU_VER == 3
-	/*
-	 * Subtract the start of domain partition regions and 1 reserved region
-	 * from total regions will get the maximum number of free regions for
-	 * memory domain partitions.
-	 */
-	return _get_num_regions() -
-	       _get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION) - 1;
-#endif
-}
-
-/**
- * @brief validate the given buffer is user accessible or not
- */
-int arc_core_mpu_buffer_validate(void *addr, size_t size, int write)
-{
-	s32_t r_index;
-
-
-	/*
-	 * For ARC MPU v2, smaller region number takes priority.
-	 * we can stop the iteration immediately once we find the
-	 * matched region that grants permission or denies access.
-	 *
-	 * For ARC MPU v3, overlapping is not supported.
-	 * we can stop the iteration immediately once we find the
-	 * matched region that grants permission or denies access.
-	 */
-#if CONFIG_ARC_MPU_VER == 2
-	for (r_index = 0; r_index < _get_num_regions(); r_index++) {
-		if (!_is_enabled_region(r_index) ||
-		    !_is_in_region(r_index, (u32_t)addr, size)) {
-			continue;
-		}
-
-		if (_is_user_accessible_region(r_index, write)) {
-			return 0;
-		} else {
-			return -EPERM;
-		}
-	}
-#elif CONFIG_ARC_MPU_VER == 3
-	r_index = _mpu_probe((u32_t)addr);
-	/*  match and the area is in one region */
-	if (r_index >= 0 && r_index == _mpu_probe((u32_t)addr + size)) {
-		if (_is_user_accessible_region(r_index, write)) {
-			return 0;
-		} else {
-			return -EPERM;
-		}
-	}
-#endif
-
-	return -EPERM;
-}
-#endif /* CONFIG_USERSPACE */
-
-/* ARC MPU Driver Initial Setup */
-
-/*
- * @brief MPU default configuration
- *
- * This function provides the default configuration mechanism for the Memory
- * Protection Unit (MPU).
- */
-static void _arc_mpu_config(void)
-{
-	u32_t num_regions;
-	u32_t i;
-
-	num_regions = _get_num_regions();
-
-	/* ARC MPU supports up to 16 Regions */
-	if (mpu_config.num_regions > num_regions) {
-		return;
-	}
-
-	/* Disable MPU */
-	arc_core_mpu_disable();
 
 #if CONFIG_ARC_MPU_VER == 2
-	u32_t r_index;
-	/*
-	 * the MPU regions are filled in the reverse order.
-	 * According to ARCv2 ISA, the MPU region with smaller
-	 * index has higher priority. The static background MPU
-	 * regions in mpu_config will be in the bottom. Then
-	 * the special type regions will be above.
-	 *
-	 */
-	r_index = num_regions - mpu_config.num_regions;
-
-	/* clear all the regions first */
-	for (i = 0U; i < r_index; i++) {
-		_region_init(i, 0, 0, 0);
-	}
-
-	/* configure the static regions */
-	for (i = 0U; i < mpu_config.num_regions; i++) {
-		_region_init(r_index,
-			     mpu_config.mpu_regions[i].base,
-			     mpu_config.mpu_regions[i].size,
-			     mpu_config.mpu_regions[i].attr);
-		r_index++;
-	}
-
-	/* default region: no read, write and execute */
-	arc_core_mpu_default(0);
-
+#include "arc_mpu_v2_internal.h"
```

Comment:
I'm not a big fan of putting full implementations  in header files and including them conditionally, instead of solving this in makefiles: compile internal_v2.c or internal_v3.c

Reply:
I agree with @ruuddw. My understanding is that this was modeled on the ARM code, @ioannisg any detail on why this approach was taken?

Reply:
One benifit is that, for ARC,  no need to change kconfig files. Current   CONFIG_ARC_MPU_VER is used to select arc mpu implementations. But it seems that  zephyr build system does not support
`zephyr_library_sources_if(CONFIG_ARC_MPU_VER == 2, arc_mpu_v2_internal.c)`
`zephyr_library_sources_if(CONFIG_ARC_MPU_VER == 3, arc_mpu_v3_internal.c)`

Reply:
ok - if the build system doesn't support checking values, this approach is for now better than changing everything into boolean CONFIG_V2, CONFIG_V3, etc.